### PR TITLE
Ignore symbols used for subsetting in `object_name_linter`

### LINKED ISSUE #582: Spurious snake_case error when using `$` in assignment
```
get_deseq_dataset.R:60:17: style: Variable and function name style should be snake_case.
        tx2gene$TXNAME <- sub("^transcript:", "", tx2gene$TXNAME)
                ^~~~~~
```

### PR DESCRIPTION
**Pull request by AshesITR**
- [x] implement fix
- [x] add regression test
- [x] add NEWS bullet

fixes #582

### CODE REVIEW

**Comment by AshesITR** [comment]
The XPath is similar to T_and_F_symbol linter, but it uses the `ancestor` axis instead of `parent` as that caused false positives.
Will check if T_and_F_symbol needs a fix as well.

File: R/object_name_linters.R [hunk 1]
```diff
@@ -29,14 +29,15 @@ object_name_linter <- function(styles = c("snake_case", "symbols")) {
...
-
-      # Or
-      " | ",
-
-      # Right hand assignments
-      "//expr[SYMBOL or STR_CONST][preceding-sibling::RIGHT_ASSIGN]/*",
+      # assignments
+      "//SYMBOL[",
+      " not(preceding-sibling::OP-DOLLAR)",
+      " and ancestor::expr[",
```
Comment (MichaelChirico) [inline R/object_name_linters.R:35]: I have a feeling `ancestor` may be too aggressive without more conditions. Running the comparison script now to try and confirm.

Comment (AshesITR) [reply R/object_name_linters.R:35]: Sounds good. I tested "//parent::expr" before, which produced false positives (detecting symbols on the other side of assigments).
"ancestor" was the only axis that passed all the tests.

**Comment by MichaelChirico** [comment]
Attaching some comparison results:

[On this branch, not `HEAD`](https://github.com/jimhester/lintr/files/6020019/765_branch_only.txt)
[On `HEAD`, not this branch](https://github.com/jimhester/lintr/files/6020020/765_head_only.txt)

**Comment by AshesITR** [comment]
The lints absent in this PR look good. All of them are either a generic defined in the package or a dollar-subset expression.

I plan to take a look at the new lints later.

**Comment by AshesITR** [comment]
Updated XPath to exclude parts of `[` and `[[` indexing expressions and added appropriate tests.

File: tests/testthat/test-object_name_linter.R [hunk 2]
```diff
@@ -102,3 +102,16 @@ test_that("dollar subsetting only lints the first expression", {
...
+
+test_that("assignment targets of compound lhs are correctly identified", {
+  linter <- object_name_linter()
+  msg <- "Variable and function name style should be snake_case or symbols."
+
+  expect_lint("good_name[badName] <- badName2", NULL, linter)
+  expect_lint("good_name[1L][badName] <- badName2", NULL, linter)
+  expect_lint("good_name[[badName]] <- badName2", NULL, linter)
+  expect_lint("good_name[[1L]][[badName]] <- badName2", NULL, linter)
+  expect_lint("good_name[[fun(badName)]] <- badName2", NULL, linter)
```
Comment (AshesITR) [inline tests/testthat/test-object_name_linter.R:114]: This case necessitates the `ancestor` axis for the `not()` expression.

**Comment by AshesITR** [comment]
Currently rerunning on all the packages with new lints from your sample, @MichaelChirico.

**Comment by AshesITR** [comment]
I checked against the follwing packages and found no erroneous added nor removed lints.
Please take another look, I think we can merge.

<details>
<summary>List of tested Packages</summary>
assertthat
bravo
broom
callr
canvasXpress
catdap
cli
clipr
ClusterBootstrap
CorrectedFDR
cpp11
crayon
crochet
DBI
DEEVD
desc
devtools
dfoptim
Dodge
epsiwal
etma
evaluate
fcros
forcats
generics
ggplot2
gramEvol
gtable
HiddenMarkov
highr
HLSM
hms
htmlwidgets
httr
idr
ineq
knitr
labeling
leontief
LogrankA
malani
mgcv
modelr
msBP
munsell
noncompliance
paramtest
peakRAM
pillar
pkgbuild
pkgload
plyr
PPQplan
processx
progress
ps
R6
rappdirs
RColorBrewer
Rcpp
RcppArmadillo
RcppEigen
readr
readxl
rematch2
remotes
reprex
Rgb
rmarkdown
rprojroot
rstudioapi
rvest
scales
selectr
seqmon
SetRank
shiny
shinyAce
space
sra
stringr
tidyselect
tidyverse
uFTIR
VecStatGraphs2D
viridisLite
waldo
withr
yarr
</details>

File: R/object_name_linters.R [hunk 3]
```diff
@@ -29,14 +29,19 @@ object_name_linter <- function(styles = c("snake_case", "symbols")) {
...
+      "  following-sibling::LEFT_ASSIGN",
+      "  or preceding-sibling::RIGHT_ASSIGN",
+      "  or following-sibling::EQ_ASSIGN",
+      " ]",
+      " and not(ancestor::expr[",
```
Comment (MichaelChirico) [inline R/object_name_linters.R:40]: as it's written, this is a totally different expression -- is there some relationship between the `LEFT_ASSIGN` (etc) ancestor and the `OP-LEFT-BRACKET` ancestor? I'm wondering if it's safer to express both conditions at once (it might not be).

Anyway I am running on a sample of 500 packages now, let's see what comes out.

Comment (AshesITR) [reply R/object_name_linters.R:40]: The problem with combining it the conditions need to be nested, i.e. there may be no such ancestor before reaching the (first) assignment ancestor.

I couldn't get it to work in a single condition but if you habe an idea for a different one, I'm happy to try it.

The tests are pretty comprehensive now. The only missing bits are nested expressions with on-the-fly assignments, but I checked a few of these interactively, e.g. `fun(BAD <- 42)`.

**Comment by MichaelChirico** [comment]
On the large sample of packages, still turning up a lot of new-not-old and old-not-new lints. Probably too big to attach here... took a random sample of 500 from each:

[main_not_pr.txt](https://github.com/jimhester/lintr/files/6047104/main_not_pr.txt)
[pr_not_main.txt](https://github.com/jimhester/lintr/files/6047105/pr_not_main.txt)

**Comment by AshesITR** [comment]
Manually inspected main_not_pr.txt.

I found one class of false negatives that need fixing:
```
'badName' <- 42
"badName" <- 42
```

Manually inspected pr_not_main.txt

I found no false positives.
The new lints are of the form `setter(badName) <- 42` or subset assignments like `badName[good_name] <- 42`.

I'll go ahead and try to fix the false negatives and add tests for them.

**Review by MichaelChirico — APPROVED** [review]
Good stuff. Glad the comparison script has been so helpful to draw out edge cases. LGTM now.
